src/client/driver/command-executors/execute-action-command.js

```diff
@@ -2,76 +2,107 @@ import hammerhead from '../deps/hammerhead';
+    var startTime = new Date();
 
-    return ensureElementExists(selector, CHECK_ELEMENT_TIMEOUT)
+    return ensureElementExists(selector, NotFoundErrorCtor)
         .then(element => {
             var checkVisibilityTimeout = CHECK_ELEMENT_TIMEOUT - (new Date() - startTime);
 
-            return ensureElementVisible(element, checkVisibilityTimeout);
-        })
-        .then(element => {
-            progressPanel.close(true);
-            return element;
-        })
-        .catch(err => {
-            progressPanel.close(false);
-            throw err;
+            return ensureElementVisible(element, checkVisibilityTimeout, IsInvisibleErrorCtor);
+        });
+}
+
+function ensureSelectors (command) {
+    if (command.type === COMMAND_TYPE.dragToElement) {
+        var sourceSelector      = () => window.eval(command.sourceSelector);
+        var destinationSelector = () => window.eval(command.destinationSelector);
+
+        return Promise
+            .all([
+                ensureElement(sourceSelector, DragSourceNotFoundError, DragSourceIsInvisibleError),
+                ensureElement(destinationSelector, DragDestinationNotFoundError, DragDestinationIsInvisibleError)
+            ])
+            .then(res => {
+                return {
+                    sourceElement:      res[0],
+                    destinationElement: res[1]
+                };
+            });
+    }
+
+    var elementSelector = () => window.eval(command.selector);
+
+    return ensureElement(elementSelector, ActionElementNotFoundError, ActionElementIsInvisibleError)
+        .then(sourceElement => {
+            return { sourceElement };
         });
 }
 
-function runAutomation (element, command) {
+function runAutomation (elements, command) {
     var automation = null;
 
     if (command.type === COMMAND_TYPE.click)
-        automation = new ClickAutomation(element, command.options);
+        automation = new ClickAutomation(elements.sourceElement, command.options);
 
     if (command.type === COMMAND_TYPE.rightClick)
-        automation = new RClickAutomation(element, command.options);
+        automation = new RClickAutomation(elements.sourceElement, command.options);
 
     if (command.type === COMMAND_TYPE.doubleClick)
-        automation = new DblClickAutomation(element, command.options);
+        automation = new DblClickAutomation(elements.sourceElement, command.options);
+
+    if (command.type === COMMAND_TYPE.drag)
+        automation = new DragToOffsetAutomation(elements.sourceElement, command.dragOffsetX, command.dragOffsetY, command.options);
+
+    if (command.type === COMMAND_TYPE.dragToElement)
+        automation = new DragToElementAutomation(elements.sourceElement, elements.destinationElement, command.options);
 
     if (command.type === COMMAND_TYPE.hover)
-        automation = new HoverAutomation(element, command.options);
+        automation = new HoverAutomation(elements.sourceElement, command.options);
```

Comment:
Refactor this so we will have function `createAutomation` which will use `switch` internally, so you will not need `automation` variable. The rest part which is `automation.run()..` can be called inline then.